compat/mingw.c

```diff
@@ -426,6 +426,67 @@ static wchar_t *normalize_ntpath(wchar_t *wbuf)
 	return wbuf;
 }
 
+static BOOL exists(const wchar_t *wpathname)
+{
+	HANDLE hnd = CreateFileW(wpathname, GENERIC_READ,
+				 FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL,
+				 OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+	if (hnd == INVALID_HANDLE_VALUE)
+		return FALSE;
+	CloseHandle(hnd);
+	return TRUE;
+}
+
+static BOOL is_exclusively_writable(const wchar_t *wpathname)
+{
+	HANDLE hnd = CreateFileW(wpathname, GENERIC_WRITE, 0, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+	if (hnd == INVALID_HANDLE_VALUE)
+		return !is_file_in_use_error(GetLastError());
+	CloseHandle(hnd);
+	return TRUE;
+}
+
+/* rename and unlink a file. This is needed for the ability to recreate a file
+ * with the same name, for example on checkout which unlinks and recreates the file */
+static int rename_and_unlink(const wchar_t *wpathname, int (*unlinker)(const wchar_t *))
+{
+	wchar_t wtemppath[MAX_LONG_PATH];
+	wchar_t *last, *original;
+	*wtemppath = 0;
+	/* leave some space for the .unlink extension */
+	wcsncat(wtemppath, wpathname, sizeof(wtemppath) - 8);
+	original = last = wtemppath + wcslen(wtemppath);
+	/* Start with file.unlink. If it exists, append X up to 12 times */
+	wcscpy(last, L".unlink");
+	last += (sizeof(L".unlink") / sizeof (wchar_t)) - 1;
+	while (exists(wtemppath)) {
```

Comment:
Here we work hard trying not to overwrite something in the work tree, but eventually give up and overwrite it anyways. How about moving files to a temporary location under .git instead?

Reply:
This is an extreme corner case, when either the file name is 4096 characters long, or all the variants of file.unlinkX{0,12} exist.

We can move it inside git, but notice that the git directory can be on another partition or network, which will be much more expensive than local rename.

Reply:
I am not sure if the git directory can be on another partition than the work tree. Git actually searches for the Git directory by moving up the directory tree from the current directory. If it crosses a file system boundary it stops saying:

```
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
```

Reply:
My point is:
- Why do we write a custom temporary file name algorithm here if we can just use the mktmp function provided by the OS?
- If we are only trying to reduce the likelihood of overwriting a work tree file, then instead of doing expensive existence checks, we can just use a longer random string in the temporary filename to the same effect.
- If we end up leaving temporary files behind, then it's better to do this in the .git directory where no-one will see it.

I thing Git LFS is a good precedence. They cache large files in the .git directory.

Reply:
One more scenario to consider:

1. Qt Creator opens and holds a file handle to foo
1. git checkout A -- foo => foo is renamed to foo.unlink and foo.unlink is removed; but an open file handle to foo.unlink still exists
1. git checkout B -- foo => foo is renamed to foo.unlink but fails since Qt Creator still holds a file handle to the removed foo.unlink file

Reply:
Regarding git on another partition, it is possible if you use .git file.

I'll change it to use mktmp. But it was you who commented before that it might overwrite a file in the repo.

Reply:
Done. system mktemp already tests for existence.